### PR TITLE
Arangodb

### PR DESCRIPTION
**Pull request by nclro**
Add ArangoDB implementation

### CODE REVIEW

**Comment by nclro** [comment]
@clun Hello, just to let you know that I created this PR for adding arangoDB support

File: ff4j-store-arangodb/src/main/java/org/ff4j/arangodb/StoreMapper.java [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+import java.util.function.Function;
+
+import static java.util.stream.Collectors.toMap;
+import static java.util.stream.Collectors.toSet;
+
+/*
+ * #%L
+ * ff4j-store-arangodb
+ * %%
+ * Copyright (C) 2013 - 2019 FF4J
```
Comment (twillouer) [inline ff4j-store-arangodb/src/main/java/org/ff4j/arangodb/StoreMapper.java]:
```suggestion
 * Copyright (C) 2013 - 2020 FF4J
```

File: ff4j-store-arangodb/pom.xml [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+        </dependency>
+        <dependency>
+            <groupId>org.assertj</groupId>
+            <artifactId>assertj-core</artifactId>
+            <scope>test</scope>
+        </dependency>
+        <dependency>
+            <groupId>org.testcontainers</groupId>
+            <artifactId>testcontainers</artifactId>
+            <version>1.12.5</version>
```
Comment (twillouer) [inline ff4j-store-arangodb/pom.xml]: 1.13.0

File: ff4j-store-arangodb/src/main/java/org/ff4j/arangodb/document/ArangoDBEvent.java [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+import com.arangodb.entity.DocumentField;
+import lombok.AllArgsConstructor;
+import lombok.Data;
+import lombok.NoArgsConstructor;
+
+/*
+ * #%L
+ * ff4j-store-arangodb
+ * %%
+ * Copyright (C) 2013 - 2019 FF4J
```
Comment (twillouer) [inline ff4j-store-arangodb/src/main/java/org/ff4j/arangodb/document/ArangoDBEvent.java]:
```suggestion
 * Copyright (C) 2013 - 2020 FF4J
```

File: ff4j-store-arangodb/src/main/java/org/ff4j/arangodb/document/ArangoDBFeature.java [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+import org.ff4j.core.FlippingStrategy;
+
+import java.util.Map;
+import java.util.Set;
+
+/*
+ * #%L
+ * ff4j-store-arangodb
+ * %%
+ * Copyright (C) 2013 - 2019 FF4J
```
Comment (twillouer) [inline ff4j-store-arangodb/src/main/java/org/ff4j/arangodb/document/ArangoDBFeature.java]:
```suggestion
 * Copyright (C) 2013 - 2020 FF4J
```

File: ff4j-store-arangodb/src/main/java/org/ff4j/arangodb/document/ArangoDBProperty.java [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+import lombok.Data;
+import lombok.NoArgsConstructor;
+
+import java.util.Set;
+
+/*
+ * #%L
+ * ff4j-store-arangodb
+ * %%
+ * Copyright (C) 2013 - 2019 FF4J
```
Comment (twillouer) [inline ff4j-store-arangodb/src/main/java/org/ff4j/arangodb/document/ArangoDBProperty.java]:
```suggestion
 * Copyright (C) 2013 - 2020 FF4J
```

File: ff4j-store-arangodb/src/main/java/org/ff4j/arangodb/store/EventStoreArangoDB.java [hunk 6]
```diff
@@ -0,0 +1,177 @@
...
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+import java.util.concurrent.TimeUnit;
+
+/*
+ * #%L
+ * ff4j-store-arangodb
+ * %%
+ * Copyright (C) 2013 - 2019 FF4J
```
Comment (twillouer) [inline ff4j-store-arangodb/src/main/java/org/ff4j/arangodb/store/EventStoreArangoDB.java]:
```suggestion
 * Copyright (C) 2013 - 2020 FF4J
```

File: ff4j-store-arangodb/src/main/java/org/ff4j/arangodb/store/EventStoreArangoDB.java [hunk 7]
```diff
@@ -0,0 +1,177 @@
...
+        ArangoDBEvent arangoDBEvent = StoreMapper.toEventStore(e);
+        return insertEvent(arangoDBEvent);
+    }
+
+    private boolean insertEvent(ArangoDBEvent event) {
+        try {
+            eventClient.insertDocument(event);
+            return true;
+        } catch (ArangoDBException e) {
+            return false;
```
Comment (twillouer) [inline ff4j-store-arangodb/src/main/java/org/ff4j/arangodb/store/EventStoreArangoDB.java]: I think you need a logger here

File: ff4j-store-arangodb/src/main/java/org/ff4j/arangodb/store/EventStoreArangoDB.java [hunk 8]
```diff
@@ -0,0 +1,177 @@
...
+    public Event getEventByUUID(final String uuid, final Long timestamp) {
+        return findEvent(uuid).orElse(null);
+    }
+
+    private Optional<Event> findEvent(final String uuid) {
+        try {
+            Optional<ArangoDBEvent> arangoDBEvent = Optional.ofNullable(eventClient.getDocument(uuid));
+            return arangoDBEvent.map(StoreMapper::fromEventStore);
+        } catch (ArangoDBException e) {
+            return Optional.empty();
```
Comment (twillouer) [inline ff4j-store-arangodb/src/main/java/org/ff4j/arangodb/store/EventStoreArangoDB.java]: same here

File: ff4j-store-arangodb/src/main/java/org/ff4j/arangodb/store/FeatureStoreArangoDB.java [hunk 9]
```diff
@@ -0,0 +1,326 @@
...
+import java.util.function.Function;
+import java.util.function.Predicate;
+
+import static java.util.stream.Collectors.*;
+
+/*
+ * #%L
+ * ff4j-store-arangodb
+ * %%
+ * Copyright (C) 2013 - 2019 FF4J
```
Comment (twillouer) [inline ff4j-store-arangodb/src/main/java/org/ff4j/arangodb/store/FeatureStoreArangoDB.java]:
```suggestion
 * Copyright (C) 2013 - 2020 FF4J
```

File: ff4j-store-arangodb/src/main/java/org/ff4j/arangodb/store/GenericArangoDBClient.java [hunk 10]
```diff
@@ -0,0 +1,146 @@
...
+
+import java.util.List;
+
+import static java.lang.String.format;
+
+/*
+ * #%L
+ * ff4j-store-arangodb
+ * %%
+ * Copyright (C) 2013 - 2019 FF4J
```
Comment (twillouer) [inline ff4j-store-arangodb/src/main/java/org/ff4j/arangodb/store/GenericArangoDBClient.java]:
```suggestion
 * Copyright (C) 2013 - 2020 FF4J
```

File: ff4j-store-arangodb/src/main/java/org/ff4j/arangodb/store/PropertyStoreArangoDB.java [hunk 11]
```diff
@@ -0,0 +1,187 @@
...
+import java.util.*;
+import java.util.function.Function;
+
+import static java.util.stream.Collectors.*;
+
+/*
+ * #%L
+ * ff4j-store-arangodb
+ * %%
+ * Copyright (C) 2013 - 2019 FF4J
```
Comment (twillouer) [inline ff4j-store-arangodb/src/main/java/org/ff4j/arangodb/store/PropertyStoreArangoDB.java]:
```suggestion
 * Copyright (C) 2013 - 2020 FF4J
```

File: ff4j-store-arangodb/src/main/java/org/ff4j/arangodb/store/PropertyStoreArangoDB.java [hunk 12]
```diff
@@ -0,0 +1,187 @@
...
+        assertPropertyExist(name);
+        return findProperty(name).orElseThrow(() -> new PropertyNotFoundException(name));
+    }
+
+    private Optional<? extends Property<?>> findProperty(final String name) {
+        try {
+            Optional<ArangoDBProperty> arangoDBProperty = Optional.ofNullable(propertyClient.getDocument(name));
+            return arangoDBProperty.map(StoreMapper::fromPropertyStore).filter(Objects::nonNull);
+        } catch (ArangoDBException e) {
+            return Optional.empty();
```
Comment (twillouer) [inline ff4j-store-arangodb/src/main/java/org/ff4j/arangodb/store/PropertyStoreArangoDB.java]: log ?

File: ff4j-store-arangodb/src/test/java/org/ff4j/arangodb/store/ArangoDBTestContainer.java [hunk 13]
```diff
@@ -0,0 +1,39 @@
+package org.ff4j.arangodb.store;
+
+import org.testcontainers.containers.GenericContainer;
+import org.testcontainers.containers.wait.strategy.Wait;
+
+/*
+ * #%L
+ * ff4j-store-arangodb
+ * %%
+ * Copyright (C) 2013 - 2019 FF4J
```
Comment (twillouer) [inline ff4j-store-arangodb/src/test/java/org/ff4j/arangodb/store/ArangoDBTestContainer.java]:
```suggestion
 * Copyright (C) 2013 - 2020 FF4J
```

File: ff4j-store-arangodb/src/test/java/org/ff4j/arangodb/store/FeatureStoreArangoDBIT.java [hunk 14]
```diff
@@ -0,0 +1,53 @@
...
+import com.arangodb.ArangoDatabase;
+import org.ff4j.core.FeatureStore;
+import org.ff4j.test.store.FeatureStoreTestSupport;
+import org.junit.ClassRule;
+
+/*
+ * #%L
+ * ff4j-store-arangodb
+ * %%
+ * Copyright (C) 2013 - 2019 FF4J
```
Comment (twillouer) [inline ff4j-store-arangodb/src/test/java/org/ff4j/arangodb/store/FeatureStoreArangoDBIT.java]:
```suggestion
 * Copyright (C) 2013 - 2020 FF4J
```

File: ff4j-store-arangodb/src/test/java/org/ff4j/arangodb/store/PropertyStoreArangoDBIT.java [hunk 15]
```diff
@@ -0,0 +1,52 @@
...
+import com.arangodb.ArangoDatabase;
+import org.ff4j.property.store.PropertyStore;
+import org.ff4j.test.propertystore.PropertyStoreTestSupport;
+import org.junit.ClassRule;
+
+/*
+ * #%L
+ * ff4j-store-arangodb
+ * %%
+ * Copyright (C) 2013 - 2019 FF4J
```
Comment (twillouer) [inline ff4j-store-arangodb/src/test/java/org/ff4j/arangodb/store/PropertyStoreArangoDBIT.java]:
```suggestion
 * Copyright (C) 2013 - 2020 FF4J
```

File: ff4j-store-arangodb/src/main/java/org/ff4j/arangodb/StoreMapper.java (hunk 1 above)
Comment (nclro) [reply ff4j-store-arangodb/src/main/java/org/ff4j/arangodb/StoreMapper.java]: done

File: ff4j-store-arangodb/src/main/java/org/ff4j/arangodb/store/EventStoreArangoDB.java (hunk 7 above)
Comment (nclro) [reply ff4j-store-arangodb/src/main/java/org/ff4j/arangodb/store/EventStoreArangoDB.java]: done

**Comment by clun** [comment]
Thank you ! Let's release :)

**Comment by clun** [comment]
Released in 1.8.4, Thank you @nclro you rock !

**Pull request merged**